src/font.cpp

```diff
@@ -158,38 +162,65 @@ namespace {
 	FontRef const ttyp0 = std::make_shared<BitmapFont>("ttyp0", &find_ttyp0_glyph);
 
 	struct ExFont : public Font {
-		ExFont();
-		Rect GetSize(std::u32string const& txt) const override;
-		BitmapRef Glyph(char32_t code) override;
+		public:
+			enum { HEIGHT = 12, WIDTH = 12 };
+			ExFont();
+			Rect GetSize(std::string const& txt) const override;
+			Rect GetSize(char32_t ch) const override;
+			GlyphRet Glyph(char32_t code) override;
+		private:
+			BitmapRef bm;
 	};
 } // anonymous namespace
 
-BitmapFont::BitmapFont(const std::string& name, BitmapFont::function_type func)
-	: Font(name, HEIGHT, false, false), func_(func) {}
+BitmapFont::BitmapFont(const std::string& name, function_type func)
+	: Font(name, HEIGHT, false, false), func(func)
+{}
 
-Rect BitmapFont::GetSize(std::u32string const& txt) const {
+Rect BitmapFont::GetSize(char32_t ch) const {
 	size_t units = 0;
-	for (char32_t c : txt) {
-		BitmapFontGlyph const* const glyph = func_(c);
-		assert(glyph);
+	if (EP_LIKELY(!std::iscntrl(static_cast<unsigned char>(ch)))) {
+		auto glyph = func(ch);
 		units += glyph->is_full? 2 : 1;
 	}
 	return Rect(0, 0, units * HALF_WIDTH, HEIGHT);
 }
 
-BitmapRef BitmapFont::Glyph(char32_t code) {
-	BitmapFontGlyph const* const glyph = func_(code);
-	assert(glyph);
-	size_t const width = glyph->is_full? FULL_WIDTH : HALF_WIDTH;
+Rect BitmapFont::GetSize(std::string const& txt) const {
+	size_t units = 0;
+	const auto* iter = txt.data();
+	const auto* end = txt.data() + txt.size();
+	while (iter != end) {
+		auto resp = Utils::UTF8Next(iter, end);
+		auto ch = resp.ch;
+		iter = resp.next;
+		if (EP_LIKELY(!std::iscntrl(static_cast<unsigned char>(resp.ch)))) {
+			auto glyph = func(ch);
+			units += glyph->is_full? 2 : 1;
+		}
+	}
+	return Rect(0, 0, units * HALF_WIDTH, HEIGHT);
+}
 
-	BitmapRef bm = Bitmap::Create(nullptr, width, HEIGHT, 0, DynamicFormat(8,8,0,8,0,8,0,8,0,PF::Alpha));
-	uint8_t* data = reinterpret_cast<uint8_t*>(bm->pixels());
-	int pitch = bm->pitch();
+Font::GlyphRet BitmapFont::Glyph(char32_t code) {
+	// FIXME: When Player #1960 is merged we can remove this lazy init.
```

Comment:
Pull 1960 is merged already

Reply:
I honestly don't remember what I had in mind here. At this point I don't want to change the PR and have to retest so I removed the comment.